Fix issues in the example configs

PerchunPak opened the pull request:
For some reason, it had multiple syntax errors and other issues, like undefined options. Would be great to add a test for sourcing all example configs, but I don't know rust

See also https://github.com/nushell/nushell/pull/14249#discussion_r1887192408

<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

CC @NotTheDr01ds

PerchunPak commented on crates/nu-utils/src/default_files/sample_config.nu:
```diff
@@ -547,17 +549,18 @@ $env.config.color_config = (dark-theme)
 #     - 'd': dimmed
 
-# foreground, background, and cursor colors are not handled by Nushell, but can be used by 
+# foreground, background, and cursor colors are not handled by Nushell, but can be used by
```
I am unsure what to do here, example configuration should work out of the box, or else it is really confusing to try different options

PerchunPak commented on crates/nu-utils/src/default_files/sample_config.nu:
```diff
@@ -727,20 +730,19 @@ $env.config.color_config.int = {||
 $env.config.color_config.float       # Float value
-$env.config.color_config.glob        # Glob value (must be declared)
+# TODO: https://github.com/nushell/nushell/issues/14600
+#$env.config.color_config.glob        # Glob value (must be declared)
```
Probably other deleted options in this list are affected by the same linked issue

NotTheDr01ds replied:
I think these are fine.  They are simply lists of settings here.  We don't need to show a "default" or example value for each one, IMHO.  The code is uncommented so that it syntax highlighted when viewing through `nu-highlight`. Ultimately, this is "documentation" that probably gets moved over to the Book itself, and these files go away.

PerchunPak replied:
Also, this kind of approach (storing documentation as default config) is not really viable for something like nixos (where configs are managed by nix itself, you don't have a default config in your disposal)

NotTheDr01ds replied:
@PerchunPak Agreed - That's why the only method for a *user* to access them is `nu config --sample`.  The end-user (and package manager) never has access to them directly.

In the past (0.100 and before) doc and defaults were mixed.  Starting in 0.101 with these changes, they no longer are.

PerchunPak replied:
So should I return all other deleted options?

NotTheDr01ds replied:
Yes, any option that is just listed by itself without an assignment should be returned. Thanks!

NotTheDr01ds commented on crates/nu-utils/src/default_files/sample_config.nu:
```diff
@@ -480,27 +487,22 @@ $env.config.menus ++= [
 $env.config.plugin_gc.default
+# true to enable stopping of inactive plugins
 $env.config.plugin_gc.default.enabled
+# how long to wait after a plugin is inactive to stop it
```
```suggestion
# How long to wait after a plugin is inactive before stopping it
```

NotTheDr01ds commented on crates/nu-utils/src/default_files/sample_config.nu:
```diff
@@ -480,27 +487,22 @@ $env.config.menus ++= [
+# Configuration for plugin garbage collection
 $env.config.plugin_gc
 $env.config.plugin_gc.default
+# true to enable stopping of inactive plugins
```
```suggestion
# true to enable stopping inactive plugins
```

NotTheDr01ds replied:
I know this was the original text that I left in from the original config - Just noticing it could be improved as long as it's getting moved :-)

PerchunPak commented on crates/nu-utils/src/default_files/sample_config.nu:
```diff
@@ -912,18 +914,18 @@ const NU_PLUGIN_DIRS = $NU_PLUGIN_DIRS ++ [($nu.default-config-dir | path join '
 use std/util "path add"
 path add "~/.local/bin"
-path add ($env.CARGO_HOME | path join "bin")
+path add ($env.HOME | path join ".cargo/bin")
```
I don't have cargo installed, so I made it to not depend on cargo

NotTheDr01ds replied:
Well, that would break on my system ;-)

It's okay to have *examples* that don't run on everyone's system.  The point here is a dynamically computed path based on an environment variable.  Otherwise, you could simply write:

```
path add "~/.cargo/bin"
```

That's the exact same example as the one above it ;-)

NotTheDr01ds commented on crates/nu-utils/src/default_files/sample_config.nu:
```diff
@@ -912,18 +914,18 @@ const NU_PLUGIN_DIRS = $NU_PLUGIN_DIRS ++ [($nu.default-config-dir | path join '
+# before your config.nu loads), the path variable is a list that can
 # be appended to using, for example:
-$env.path ++= "~/.local/bin"
+$env.PATH ++ [ "~/.local/bin" ]
```
```suggestion
$env.PATH ++= [ "~/.local/bin" ]
```

Right?

NotTheDr01ds replied:
These worked when I wrote them.  Then the `++` and `++=` operators were changed ;-).

PerchunPak replied:
Huh, it used to raise a syntax error but no longer

PerchunPak commented on crates/nu-utils/src/default_files/sample_config.nu:
```diff
@@ -912,18 +914,18 @@ const NU_PLUGIN_DIRS = $NU_PLUGIN_DIRS ++ [($nu.default-config-dir | path join '
+# before your config.nu loads), the path variable is a list that can
 # be appended to using, for example:
-$env.path ++= "~/.local/bin"
+$env.PATH ++ [ "~/.local/bin" ]
```
I am unsure why this raises an error, while https://github.com/nushell/nushell/blob/39770d41975a0dae3cdc89802215416a4c6671e4/crates/nu-utils/src/default_files/sample_config.nu#L420 doesn't

NotTheDr01ds replied:
It's the recent change from `++` acting as an `append` to having it be a concatenate.  Before, the operand could be any value; now it must be a list.

NotTheDr01ds commented on crates/nu-utils/src/default_files/scaffold_env.nu:
```diff
@@ -1,5 +1,7 @@
 # env.nu
 #
+# version = "0.100.1"
```
There's probably a buildscript somewhere that needs to be updated in order to do this.  Otherwise it's always going to be out-of-date.

When I get back to my desk I'll try to figure this out.

NotTheDr01ds replied:
But I'm not sure we should put a version in a file that simply has *comments* anyway :-)

NotTheDr01ds commented on crates/nu-utils/src/default_files/sample_config.nu:
```diff
@@ -402,12 +400,19 @@ $env.config.ls = true
-$env.config.hooks.display_output = ""      # Before Nushell output is displayed in the terminal
-$env.config.hooks.command_not_found = []   # When a command is not found
+# Before each prompt is displayed
+$env.config.hooks.pre_prompt = [{ null }]
```
I can understand moving the comments for consistency, but why the `null` additions?

PerchunPak replied:
Those were in 0.100.0 example config, and because `$env.config.hooks.env_change` had wrong type, I decided to return default value as an example. I can remove all nulls if you prefer so

NotTheDr01ds reviewed (commented):
I think it would be good to have a PR that we can quickly approve with the typo fixes.  We can sort the others out separately, but I'd like to make sure your typo corrections get into 0.101.

NotTheDr01ds commented:
I think we should add something along the lines of:

```
# This file is not intended to be run as-is in Nushell. It is for documentation purposes only. Running this file *will* generate errors.
```

PerchunPak commented:
Shouldn't `NU_LIB_DIRS` and `NU_PLUGIN_DIRS` be in env.nu?

NotTheDr01ds commented:
> Shouldn't `NU_LIB_DIRS` and `NU_PLUGIN_DIRS` be in env.nu?

Not any longer.  As primarily constants, they can be set and used in the same file.  There's really nothing that we have to set in `env.nu` any longer, so we just moved the entire "sample" over to `sample_config.nu`.

PerchunPak commented:
> Not any longer. As primarily constants now, they can be set and used in the same file. There's really nothing that we have to set in env.nu any longer, so we just moved the entire "sample" over to sample_config.nu.

For me, they didn't apply inside config.nu until I moved them to env.nu (`source nu_scripts/...` didn't work inside config.nu but did work in the interactive shell)

fdncred commented:
> For me, they didn't apply inside config.nu until I moved them to env.nu (`source nu_scripts/...` didn't work inside config.nu but did work in the interactive shell)

They work for me on the latest main. I just tested again.

fdncred commented:
I also think all config files should have a version number to show which version of nushell they're compatible with.

fdncred reviewed (changes requested):
If this is a "sample" people will put it in place as the config and try it out. This file should work. I've added a few comments where it doesn't work but there are more than I listed.

Release is quickly approaching and we need to get this resolved.

NotTheDr01ds commented:
> If this is a "sample" people will put it in place as the config and try it out. This file should work. I've added a few comments where it doesn't work but there are more than I listed.
> 
> Release is quickly approaching and we need to get this resolved.

Ack - No.  This is *documentation*.  This was *never* designed to be run.  I'll change the name of the file and command to be more clear on that.

fdncred commented:
> > If this is a "sample" people will put it in place as the config and try it out. This file should work. I've added a few comments where it doesn't work but there are more than I listed.
> > Release is quickly approaching and we need to get this resolved.
> 
> Ack - No. This is _documentation_. This was _never_ designed to be run. I'll change the name of the file and command to be more clear on that.

If we don't want people trying to use these files, which seems like they intuitively would, given the file names, then the files should have documentation (or something other than sample/example) in the name and the first lines of the file should state that it's documentation and not intended to be used as a configuration file. If they're going to be documentation, hopefully they'll be complete, otherwise I'm not sure what purpose they serve. e.g. I'm not sure how helpful it is to say "here's _some_ of the settings you can use" vs "this is a list of all settings and what they mean".

PerchunPak commented:
@NotTheDr01ds should be done, except for https://github.com/nushell/nushell/pull/14601#issuecomment-2548901573 (but I feel that https://github.com/nushell/nushell/pull/14608 is more suitable for this change)

NotTheDr01ds commented:
@fdncred I think the last commit gets the actual *errors* that were in the documentation resolved.

If you are ready to merge it, I'll rebase #14608 on it so that we reduce (hopefully eliminate) any confusion on how these files are to be used.

fdncred commented:
ok, thanks.
